ibis/backends/duckdb/__init__.py

```diff
@@ -1145,57 +1148,154 @@ def to_xlsx(
+        more_options: Iterable[str] = [],
     ) -> None:
-        """Attach another DuckDB database to the current DuckDB session.
+        """Attach a DuckDB, sqlite, postgres, or mysql database to the current DuckDB session.
+
+        See [DuckDB documentation](https://duckdb.org/docs/stable/sql/statements/attach.html)
+        for more information.
 
         Parameters
         ----------
-        path
-            Path to the database to attach.
+        path_or_url
+            Path or url to the database to attach.
         name
             Name to attach the database as. Defaults to the basename of `path`.
+        on_exists
+            What to do if the database already exists.
+            If set to `"error"`, raise an error if the database already exists.
+            If set to `"ignore"`, do nothing if the database already exists.
+            If set to `"replace"`, detach the existing database and attach the new one.
         read_only
             Whether to attach the database as read-only.
+        type
+            The type of the database to attach. If none, infers from the file extension.
+        more_options
+            Additional options, e.g. `["STORAGE_VERSION 'v1.2.0'", "block_size 262144"]`.
 
+        Returns
+        -------
+        None | str
+            The name of the attached database, or `None` if the database already exists.
         """
-        code = f"ATTACH '{path}'"
+        if on_exists == "ignore":
+            on_exists_string = "IF NOT EXISTS"
+        elif on_exists == "replace":
+            on_exists_string = "OR REPLACE"
+        elif on_exists == "error":
+            on_exists_string = ""
+        else:
+            raise ValueError(
+                f"Invalid value for `on_exists`: {on_exists!r}. "
+                "Must be one of 'error' or 'ignore'."
+            )
 
-        if name is not None:
-            name = sg.to_identifier(name).sql(self.name)
-            code += f" AS {name}"
+        if name is None:
+            name = _attach_name(path_or_url)
 
-        if read_only:
-            code += " (READ_ONLY)"
+        as_name = f"AS {sg.to_identifier(name, self.compiler.quoted).sql(self.name)}"
 
-        self.con.execute(code).fetchall()
+        options = [*more_options]
+        if read_only:
+            options.append("READ_ONLY true")
+        else:
+            options.append("READ_ONLY false")
+        if type is not None:
+            options.append(f"TYPE {type.upper()}")
+        option_string = ", ".join(options)
+
+        sql = f"ATTACH {on_exists_string} '{path_or_url}' {as_name} ({option_string})"
+        databases_before = set(self.list_catalogs())
+        self.con.execute(sql).fetchall()
+        if name in databases_before:
+            if on_exists == "ignore":
+                return None
+            if on_exists == "replace":
+                return name
+            raise AssertionError(
```

Comment:
This looks like a code path that a user can hit. Why raise an `AssertionError` here instead of something more bespoke?